### ISSUE TITLE
08-k-nearest-neighbours-winequality-red.ipynb

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/AnthonnyMal-Machine-Learnig#7]
Hola @AnthonnyMal , por aquí las issues del ejercicio de clustering!

## LO QUE HAS HECHO BIEN

- **Selección de variables correcta**: MedInc, Latitude, Longitude
- **Eliminación de duplicados**: Verificas antes y después con `duplicated().sum()`
- **EDA con histogramas**: Distribución de cada variable
- **Scatter plot geográfico inicial**: Buena visualización previa al clustering
- **MinMaxScaler**: Escalado correcto para K-Means
- **KElbowVisualizer de Yellowbrick**: Excelente herramienta para el método del codo
- **Visualización de clusters por coordenadas**: Scatter plot con colorbar
- **Histograma de MedInc por cluster**: Muy buena visualización para entender la segmentación por ingresos
- **Observaciones interpretativas**: Buena explicación de cómo el modelo agrupa por ingresos y ubicación
- **DecisionTreeClassifier con max_depth=5**: Limitas la profundidad para evitar overfitting
- **Verificación train y test accuracy**: Compruebas ambos para detectar overfitting

---

### 1. ERROR GRAVE: Escalas TODO el dataset ANTES del split (DATA LEAKAGE)

```python
min_max_scaler = MinMaxScaler()
min_max_features = min_max_scaler.fit_transform(df)  # ← Escalas TODO
df_escalado = pd.DataFrame(min_max_features, ...)
```

El scaler "ve" todos los datos (incluyendo lo que luego será test) al calcular min/max. Esto es **data leakage**.

**La forma correcta:**
```python
# 1. Split PRIMERO (sobre datos sin escalar)
X_train, X_test = train_test_split(df, test_size=0.2, random_state=18)

# 2. Escalar DESPUÉS
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)  # fit_transform en train
X_test_scaled = scaler.transform(X_test)        # solo transform en test

# 3. K-Means en train
kmeans = KMeans(n_clusters=6, random_state=18)
kmeans.fit(X_train_scaled)

# 4. Predecir clusters en test
train_clusters = kmeans.labels_
test_clusters = kmeans.predict(X_test_scaled)
```

---

### 2. OPCIONAL: Falta Silhouette Score

Lo importas pero nunca lo usas, es una métrica muy usada con este algoritmo :) :

```python
from sklearn.metrics import silhouette_score  # ← Importado pero no usado
```

Añade:
```python
silhouette = silhouette_score(df_escalado.drop('cluster', axis=1), df_escalado['cluster'])
print(f"Silhouette Score: {silhouette:.4f}")
```
---

### 3. Falta matriz de confusión para el clasificador.

Intenta recordar que este gráfico es crucial en casi todos los proyectos :) 

```python
from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_pred_test)
plt.figure(figsize=(8, 6))
sns.heatmap(cm, annot=True, fmt='d', cmap='Blues',
            xticklabels=[f'Cluster {i}' for i in range(6)],
            yticklabels=[f'Cluster {i}' for i in range(6)])
plt.xlabel('Predicción')
plt.ylabel('Real')
plt.title('Matriz de Confusión - DecisionTree')
plt.show()
```

---

### 4. Falta classification_report

Vamos a empezar a jugar a intentar ver todas las métricas ! 

```python
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred_test))
```

---

### 5. Leve. Falta guardar los modelos

```python
import pickle

# Guardar K-Means
with open("../models/kmeans_housing.pkl", "wb") as f:
    pickle.dump(kmeans, f)

# Guardar Scaler (IMPORTANTE)
with open("../models/scaler_housing.pkl", "wb") as f:
    pickle.dump(min_max_scaler, f)

# Guardar DecisionTree
with open("../models/decision_tree_clusters.pkl", "wb") as f:
    pickle.dump(model, f)
```

---

**Anthonny**, el ejercicio tiene buenas visualizaciones y me gusta especialmente el histograma de MedInc por cluster - demuestra cómo K-Means segmenta por ingresos. El uso de KElbowVisualizer de Yellowbrick también es un plus. Sin embargo, el error de **data leakage** al escalar antes del split debes corregirlo. También te falta guardar los modelos y el scaler (sin el scaler no podrás predecir nuevos datos correctamente). Importas `silhouette_score` pero no lo usas - añádelo para evaluar la calidad del clustering.
